src/plugins/telegram/index.ts

```diff
@@ -1,45 +1,70 @@
 import axios from 'axios';
+import IPlugin, { DefaultPlugin } from '../Plugin';
 
-// TODO: harmonize plugin interface
-// TODO: .env?
-const API_URL = `https://api.telegram.org/bot${process.env.TELEGRAM_TOKEN}`;
+type TelegramGetUpdateResponse = {
+  result: [{ message: { from: { username: string; id: string } } }];
+};
+type TelegramSendMessageResponse = {
+  result: { message: { message_id: string } };
+};
 
-// Gets user id from username
-// This function can be optimized by storing id's inside the DB on first notifications
-async function getUserId(handles: string[]): Promise<string[]> {
-  const UserIds: string[] = [];
+class TelegramPlugin extends DefaultPlugin implements IPlugin {
+  #BASE_URL;
 
-  try {
-    const response = await axios.get(`${API_URL}/getUpdates`);
+  constructor(telegramToken: string) {
+    super(TelegramPlugin.name);
+    this.#BASE_URL = `https://api.telegram.org/bot${telegramToken}`;
+  }
+
+  async #getUserId(targets: string[]): Promise<string[]> {
+    // this shit load all the conversations of the bot
+    // meaning chats contains all the conversations started with the bot
+    // that include username of all of our users
+    // TODO: find a better way to handle this
+    const response = await axios.get<TelegramGetUpdateResponse>(
+      `${this.#BASE_URL}/getUpdates`
+    );
     const chats = response.data.result;
-    if (chats.length > 0) {
-      for (let i = 0; i < chats.length; i++) {
-        if (chats[i].message.from.username == handles[0])
-          UserIds.push(chats[i].message.from.id);
-      }
-    } else {
-      throw new Error('➡️ User has not started conversation with user');
-    }
-  } catch (error) {
-    console.error(error);
+
+    // get the user id of the targetted users
+    const UserIds = chats.reduce((acc: string[], curr) => {
+      const isIncluded = targets.includes(curr.message.from.username);
+      return isIncluded ? [...acc, curr.message.from.id] : acc;
+    }, []);
+
+    // throw an error if users cannot be found (meaning targetted users didn't start a conversation with the bot)
+    // *OPINIONATED*: we only throw an error if all of the targetted users has not started a conversation with the bot
+    if (UserIds.length === 0)
+      this.throwError('➡️ User has not started conversation with user');
+
+    return UserIds;
   }
-  return UserIds;
-}
 
-// Send the notification message from user chat Id
-export async function notify(message: string, handles: string[]) {
-  const chatId = await getUserId(handles);
-  for (let i = 0; i < chatId.length; i++) {
-    try {
-      const res = await axios.post(`${API_URL}/sendMessage`, {
-        chat_id: chatId[i],
+  async notify(message: string, targets: string[]): Promise<void> {
+    const endpoint = `${this.#BASE_URL}/sendMessage`;
+    const chatId = await this.#getUserId(targets);
```

Comment:
I would rename `chatId` to `chatIds` to avoid confusion since this variable represents a collection (list) of chat ids

```suggestion
    const chatIds = await this.#getUserId(targets);
```